tests/unit/test_charm.py

```diff
@@ -1,68 +1,256 @@
-# Copyright 2024 Ali
+# Copyright 2024 Canonical Ltd.
 # See LICENSE file for licensing details.
 #
 # Learn more about testing at: https://juju.is/docs/sdk/testing
 
-import unittest
 
-import ops
-import ops.testing
-from charm import AirbyteUiK8SOperatorCharm
+"""Charm unit tests."""
 
+# pylint:disable=protected-access
+
+from unittest import TestCase, mock
+
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
+from ops.pebble import CheckStatus
+from ops.testing import Harness
+
+from charm import AirbyteUIK8sOperatorCharm
+from src.charm import CONNECTOR_BUILDER_API_PORT, INTERNAL_API_PORT, WEB_UI_PORT
+
+APP_NAME = "airbyte-webapp"
+mock_incomplete_pebble_plan = {"services": {"airbyte-webapp": {"override": "replace"}}}
+
+
+class TestCharm(TestCase):
+    """Unit tests for charm.
+
+    Attrs:
+        maxDiff: Specifies max difference shown by failed tests.
+    """
+
+    maxDiff = None
 
-class TestCharm(unittest.TestCase):
     def setUp(self):
-        self.harness = ops.testing.Harness(AirbyteUiK8SOperatorCharm)
+        """Create setup for the unit tests."""
+        self.harness = Harness(AirbyteUIK8sOperatorCharm)
         self.addCleanup(self.harness.cleanup)
+        self.harness.set_can_connect(APP_NAME, True)
+        self.harness.set_leader(True)
+        self.harness.set_model_name("airbyte-model")
         self.harness.begin()
 
-    def test_httpbin_pebble_ready(self):
-        # Expected plan after Pebble ready with default config
-        expected_plan = {
+    def test_initial_plan(self):
+        """The initial pebble plan is empty."""
+        initial_plan = self.harness.get_container_pebble_plan(APP_NAME).to_dict()
+        self.assertEqual(initial_plan, {})
+
+    def test_blocked_by_peer_relation_not_ready(self):
+        """The charm is blocked without a peer relation."""
+        harness = self.harness
+
+        # Simulate pebble readiness.
+        container = harness.model.unit.get_container(APP_NAME)
+        harness.charm.on.airbyte_webapp_pebble_ready.emit(container)
+
+        # No plans are set yet.
+        got_plan = harness.get_container_pebble_plan(APP_NAME).to_dict()
+        self.assertEqual(got_plan, {})
+
+        # The BlockStatus is set with a message.
+        self.assertEqual(harness.model.unit.status, BlockedStatus("peer relation not ready"))
+
+    def test_ingress(self):
```

Comment:
It could make sense to split these 3 ingress assertions into separate tests so it's more clear which one is failing.